### ISSUE TITLE
Blob Snapshots... Awkward

### ISSUE THREAD

**Issue by nbering** [Azure/azure-storage-node#123]
I like the idea of blob snapshots, I'm using them to archive and manage old versions of a backup. It's nice because it makes the most recent version very easy to access, while hiding the potentially large list of old backups.

I've tagged my snapshots with an expires date using metadata and whenever I create a new backup, I iterate through the old snapshots and delete any that are past their expiry date.

The problem I have is that the API doesn't work very well for managing snapshots on just one blob. To get snapshots I have to search the entire container and then filter for the blob I'm looking for. I think I could probably refine this a bit by using `listBlobsSegmentedWithPrefix()`, but I feel that since the snapshot pertains to a particular blob, we should be able to enumerate snapshots for a blob. Perhaps using the `getBlobProperties()` method should include an array of snapshotId's?

I suspect this awkwardness comes from the upstream Azure Storage REST API. If you could point me in the right direction I could submit this feedback to the team responsible for that API.

I'm also guessing that this awkwardness with the API is why it seems the Azure-CLI project contributors seem hesitant to add snapshot support to their tooling. It's been brought up and then closed, then brought up again and there doesn't seem to be any commitment to implement snapshot support. This puts cross-platform users into an awkward situation because the tools don't support snapshots, but snapshots can block users from deleting blobs.

A [blog post](https://www.simple-talk.com/cloud/platform-as-a-service/azure-blob-storage-part-7-snapshots/) I found suggests that the .Net API does have the ability to enumerate snapshots on a per-blob basis, but I could be incorrect in my interpretation.

> **TL;DR**: It would be nice to be able to get a list of snapshots, per-blob, without worrying about the rest of the container.

**Comment by yaxia** [comment]
Could you please try the `BlobService.listBlobsSegmented` API, with the blob name set as the `prefix` argument and `options.include = 'snapshot'`?

```javascript

var blobService = /* create the service client*/
var options = { include: 'snapshots' };
blobService.listBlobsSegmentedWithPrefix(contianerName, blobName, null, options, function(error, results) {
  ...
});

```

**Comment by nbering** [comment]
I've done this, it's just awkward because you're operating against the whole container, not a specific blob. Adding the prefix filers it down considerably, but depending on naming conventions it's still possible to have results from separate blobs returned, requiring the developer to filter them out.

Like I mentioned, it's not impossible, it just makes snapshots feel like an after-thought. Also, because it's not easily discoverable, it made something that seemed like it was going to be a couple of hours take most of the day to figure out. That's the worst kind of API quirk.

**Comment by vinjiang** [comment]
@nbering what .Net provides is as same as Node.js. This is align with the REST api provided by service. We're going to close this issue and feel free to reopen it if you have further questions/concerns.

**Comment by nbering** [comment]
@vinjiang Thanks for setting me straight on that point. I looked into the implementation details on the .Net side and realized that this is in fact the case.
